pkg/obfuscate/shell_scanner.go

```diff
@@ -0,0 +1,56 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package obfuscate
+
+import (
+	"regexp"
+)
+
+type ShellScanner struct {
+	s     string
+	index int
+	match *Match
+}
+
+type Match struct {
+	s string
+}
+
+func NewShellScanner(input string) *ShellScanner {
+	return &ShellScanner{s: input, index: 0, match: nil}
+}
+
+func (s *ShellScanner) String() string {
+	return s.s
+}
+
+func (s *ShellScanner) SetIndex(newIndex int) {
+	s.index = newIndex
+}
+
+func (s *ShellScanner) Index() int {
+	return s.index
+}
+
+func (m *Match) String() string {
+	return m.s
+}
+
+func (s *ShellScanner) Scan(re *regexp.Regexp) *Match {
+	loc := re.FindStringIndex(s.s[s.index:])
+	if loc == nil {
+		return nil
+	}
+	s.match = &Match{s: s.s[s.index+loc[0] : s.index+loc[1]]}
+
+	// Return nil if the match doesn't start at the current index
+	if loc[0] != 0 {
```

Comment:
That feels extremely inefficient 🤔 Worth prepending every regex with `^\s*` ?

Reply:
Even with adding `^\s*` at the start of every regex (and by refactoring all code)
The `redirection` regex `^\s*(&>)|(([0-9])?((>\|)|([<>]&[0-9\-]?)|([<>]?>)|(<(<(<)?(-)?)?)))` (with the prepending) will still match even if it's not the start of the string and ending in the if condition.

I don't know which other regex can be used to do something similar without the need of this if condition. But adding `^\s*` to other regexes can still have been an improvement. I committed the code as is with this only improvement.

Reply:
If the regex you just shared, `^\s*` was considered as part of the `(&>)` alternative pattern. You want to put everything within parenthesis as follow: `^\s*(?:(&>)|(([0-9])?((>\|)|([<>]&[0-9\-]?)|([<>]?>)|(<(<(<)?(-)?)?))))` :)